File: pyrithmetic/Subtraction.py
import random

from .Task import Task


class SubtractionTask(Task):
    def __init__(
        self, id, *, difficulty_level, accuracy_target=0.80, solve_time_target=3
    ):
        super().__init__(
            id,
            accuracy_target=accuracy_target,
            solve_time_target=solve_time_target,
            difficulty_level=difficulty_level,
        )
        self.operator = "-"

    def generate_assignment(self):
        raise NotImplementedError("Subclasses must implement generate_assignment()")


class SubtractFrom10(SubtractionTask):
    def __init__(self, difficulty_level=20):
        super().__init__("10-5", difficulty_level=difficulty_level)

    def generate_assignment(self):
        num1 = 10
        num2 = random.randint(1, 9)
        task = f"{num1} {self.operator} {num2}"
        correct_answer = num1 - num2
        return task, correct_answer


class SubtractSingleDigitFromTwoDigitRoundTask(SubtractionTask):
    def __init__(self, difficulty_level=30):
        super().__init__("37-7", difficulty_level=difficulty_level)

    def generate_assignment(self):
        num2 = random.randint(1, 9)
        num1 = num2 + random.choice([20, 30, 40, 50, 60, 70, 80, 90])
        task = f"{num1} {self.operator} {num2}"
        correct_answer = num1 - num2
        return task, correct_answer


class SubtractSingleDigitFromTweenTask(SubtractionTask):
    def __init__(self, difficulty_level=50):
        super().__init__("15-8", difficulty_level=difficulty_level)

    def generate_assignment(self):
        num2 = random.randint(1, 9)
        num1 = random.randint(0, num2 - 1)
        num1 += 10
        task = f"{num1} {self.operator} {num2}"
        correct_answer = num1 - num2
        return task, correct_answer


class SubtractSingleDigitFromTwoDigitTask(SubtractionTask):
    def __init__(self, difficulty_level=60):
        super().__init__("25-8", difficulty_level=difficulty_level)

    def generate_assignment(self):
        num2 = random.randint(1, 9)
        num1 = random.randint(0, num2 - 1)
        num1 += random.choice([10, 20, 30, 40, 50, 60, 70, 80, 90])
        task = f"{num1} {self.operator} {num2}"
        correct_answer = num1 - num2
        return task, correct_answer


class SubtractNoCarryTask(SubtractionTask):
    def __init__(self, difficulty_level, num_digits=1, second_digits=None):
        if second_digits is None:
            second_digits = num_digits
        if second_digits < num_digits:
            num_digits, second_digits = second_digits, num_digits
        self.first_digits = num_digits
        self.second_digits = second_digits
        id = f"{'5'*self.second_digits}-{'2'*self.first_digits}"
        super().__init__(id, difficulty_level=difficulty_level)

    def generate_number_pair(self, include_zeros=False):
        if include_zeros:
            num1 = random.randint(0, 9)
            num2 = random.randint(0, num1)
        else:
            num1 = random.randint(1, 9)
            num2 = random.randint(1, num1)
        return num1, num2

    def generate_assignment(self):
        num1 = 0
        num2 = 0
        for i in range(self.first_digits):
            include_zeros = not (i == 0 or i == self.first_digits - 1)
            a, b = self.generate_number_pair(include_zeros=include_zeros)
            num1 += a * 10**i
            num2 += b * 10**i
        for i in range(self.first_digits, self.second_digits):
            num1 += random.randint(1, 9) * 10**i
        task = f"{num1} {self.operator} {num2}"
        correct_answer = num1 - num2
        return task, correct_answer


class SubtractTwoDigitFromTwoDigitTask(SubtractionTask):
    def __init__(self, difficulty_level=80):
        super().__init__("54-38", difficulty_level=difficulty_level)

    def generate_assignment(self):
        num1 = random.randint(11, 99)
        num2 = random.randint(11, num1)
        task = f"{num1} {self.operator} {num2}"
        correct_answer = num1 - num2
        return task, correct_answer


class SubtractTwoDigitFromThreeDigitTask(SubtractionTask):
    def __init__(self, difficulty_level=100):
        super().__init__("254-38", difficulty_level=difficulty_level)

    def generate_assignment(self):
        num1 = random.randint(11, 99)
        num2 = random.randint(1, num1)
        num1 += random.choice([100, 200, 300, 400, 500, 600, 700, 800, 900])
        task = f"{num1} {self.operator} {num2}"
        correct_answer = num1 - num2
        return task, correct_answer


class SubtractTwoDigitFromThreeDigitTaskWithCarry(SubtractionTask):
    def __init__(self, difficulty_level=110):
        super().__init__("254-68", difficulty_level=difficulty_level)

    def generate_assignment(self):
        num1 = random.randint(101, 999)
        num2 = random.randint(1, 99)
        task = f"{num1} {self.operator} {num2}"
        correct_answer = num1 - num2
        return task, correct_answer


class SubtractThreeDigitFromThreeDigitTask(SubtractionTask):
    def __init__(self, difficulty_level=120):
        super().__init__("254-168", difficulty_level=difficulty_level)

    def generate_assignment(self):
        num1 = random.randint(101, 999)
        num2 = random.randint(101, num1)
        task = f"{num1} {self.operator} {num2}"
        correct_answer = num1 - num2
        return task, correct_answer

File: js/Subtraction.js
import * as Task from "./Task.js";
import * as utils from "./utils.js";

class SubtractionTask extends Task.Task {
  constructor(id, { accuracyTarget = 0.8, solveTimeTarget = 3 }) {
    super(id, { accuracyTarget, solveTimeTarget });
    this.operator = "-";
  }

  generateAssignment() {
    throw new Error("Subclasses must implement generateAssignment()");
  }
}

export class SubtractFrom10 extends SubtractionTask {
  constructor() {
    super("10-5", {});
  }

  generateAssignment() {
    const num1 = 10;
    const num2 = utils.randint(1, 9);
    const task = `${num1} ${this.operator} ${num2}`;
    const correctAnswer = num1 - num2;
    return { task, correctAnswer };
  }
}

export class SubtractSingleDigitFromTwoDigitRoundTask extends SubtractionTask {
  constructor() {
    super("37-7", {});
  }

  generateAssignment() {
    const num2 = utils.randint(1, 9);
    const num1 = num2 + utils.randint(2, 9) * 10;
    const task = `${num1} ${this.operator} ${num2}`;
    const correctAnswer = num1 - num2;
    return { task, correctAnswer };
  }
}

export class SubtractSingleDigitFromTweenTask extends SubtractionTask {
  constructor() {
    super("15-8", {});
  }

  generateAssignment() {
    const num2 = utils.randint(1, 9);
    const num1 = utils.randint(0, num2 - 1) + 10;
    const task = `${num1} ${this.operator} ${num2}`;
    const correctAnswer = num1 - num2;
    return { task, correctAnswer };
  }
}

export class SubtractSingleDigitFromTwoDigitTask extends SubtractionTask {
  constructor() {
    super("25-8", {});
  }

  generateAssignment() {
    const num2 = utils.randint(1, 9);
    const num1 = utils.randint(1, num2 - 1);
    num1 += utils.randint(1, 9) * 10;
    const task = `${num1} ${this.operator} ${num2}`;
    const correctAnswer = num1 - num2;
    return { task, correctAnswer };
  }
}

export class SubtractNoCarryTask extends SubtractionTask {
  constructor(num_digits = 1, second_digits = null) {
    if (second_digits === null) {
      second_digits = num_digits;
    }
    if (second_digits <= num_digits) {
      [num_digits, second_digits] = [second_digits, num_digits];
    }
    let id = `${"5".repeat(second_digits)}-${"2".repeat(num_digits)}`;
    super(id, {});
    this.first_digits = num_digits;
    this.second_digits = second_digits;
  }

  generateNumberPair(include_zeros = false) {
    let num1 = 0;
    let num2 = 0;
    if (include_zeros) {
      num1 = utils.randint(0, 9);
      num2 = utils.randint(0, num1);
    } else {
      num1 = utils.randint(1, 9);
      num2 = utils.randint(1, num1);
    }
    return [num1, num2];
  }

  generateAssignment() {
    let num1 = 0;
    let num2 = 0;

    for (let i = 0; i < this.first_digits; i++) {
      let includeZeros = !(i === 0 || i === this.first_digits - 1);
      let [a, b] = this.generateNumberPair(includeZeros);
      num1 += a * Math.pow(10, i);
      num2 += b * Math.pow(10, i);
    }

    for (let i = this.firstDigits; i < this.secondDigits; i++) {
      num1 += utils.randint(1, 9) * Math.pow(10, i);
    }

    let task = `${num1} ${this.operator} ${num2}`;
    let correctAnswer = num1 - num2;

    return { task, correctAnswer };
  }
}

export class SubtractTwoDigitFromTwoDigitTask extends SubtractionTask {
  constructor() {
    super("54-38", {});
  }

  generateAssignment() {
    const num1 = utils.randint(11, 99);
    const num2 = utils.randint(11, num1);
    const task = `${num1} ${this.operator} ${num2}`;
    const correctAnswer = num1 - num2;
    return { task, correctAnswer };
  }
}

export class SubtractTwoDigitFromThreeDigitTask extends SubtractionTask {
  constructor() {
    super("254-38", {});
  }

  generateAssignment() {
    const num1 = utils.randint(11, 99);
    const num2 = utils.randint(11, num1);
    num1 += utils.randint(1, 9) * 100;
    const task = `${num1} ${this.operator} ${num2}`;
    const correctAnswer = num1 - num2;
    return { task, correctAnswer };
  }
}

export class SubtractTwoDigitFromThreeDigitTaskWithCarry extends SubtractionTask {
  constructor() {
    super("254-68", {});
  }

  generateAssignment() {
    const num1 = utils.randint(101, 99);
    const num2 = utils.randint(1, 99);
    const task = `${num1} ${this.operator} ${num2}`;
    const correctAnswer = num1 - num2;
    return { task, correctAnswer };
  }
}

export class SubtractThreeDigitFromThreeDigitTask extends SubtractionTask {
  constructor() {
    super("254-168", {});
  }

  generateAssignment() {
    const num1 = utils.randint(101, 99);
    const num2 = utils.randint(101, num1);
    const task = `${num1} ${this.operator} ${num2}`;
    const correctAnswer = num1 - num2;
    return { task, correctAnswer };
  }
}
